File: 05_mapreduce_python/mapreduce.py
import sys;
from itertools import groupby;

#
# Pyton supports MapReduce paradigm (and more elaborate functional programming)
# by the following functions:
#
# map (function,list) applies the given function over the collection
# filter(function, collection) filters using the given function
# reduce (function, collection) applies reduce successively over the list (slightly different from MapReduce reduce)

# This makes a pure python wordcount mapreduce very easy:


# Now, words is a list of all the words from input.txt. Let us map it

def mapfunc(w):
    # Let us remove all puncuation and spaces.  
    cleanword = ''.join([i for i in w if i.isalpha()])
    return [cleanword,1];

def reducefunc(key, values):
    counts = [x[1] for x in values];
    return [key,sum(counts)];

# MapReduce performs the following:
# Step 1: Read the input and split it for the mapper, here we split already into words
with open("input.txt") as f:
    words=[word for line in f for word in line.split()]
# Map each of those, here we use a mapper that removes all non-alpha characters and emits pairs.
map_result = map (mapfunc, words)
# Sort the mapped results by using the first (e.g. the word)
map_result_sorted = sorted (map_result, key = lambda x: x[0])
# Now apply the reducer for each group of equal keys
reduce_result = [];
for k, g in groupby(map_result_sorted, key = lambda x: x[0]):
    reduce_result.append(reducefunc(k, list(g)))
# print out the results:
print reduce_result

File: 05_mapreduce_python/mapreduce.R
#
# This R sample shall get you started with mapreduce in R
# Note that we follow some naming such that you can actually
# run mapreduce algorithms developed in this environment
# over Hadoop using, for example, rhadoop.
#
#
# Look at the commented lines at the end of this file for a real-world
# MapReduce over Hadoop, which first distributes data to the distributed file sysmte (DFS)
# and then mapreduces using the very same functions.
#
# you might need to run install.packages("plyr") before using this, as the PLYR package is
# used for grouping

library(plyr);




# Step 1: Read lines
lines = readLines(file("input.txt","r"));

# first, remove all non-alphabetic chars (gsub), then split at space
words = strsplit(gsub("[^A-Za-z ]","",lines),' ');
words = as.list(unlist(words))

mapfunc <- function(k)
{
   list(key = k,value=1);
}



# Step 1: Map all words to mappers
mapresult = lapply(words, mapfunc) 
# Step 2: Sort using key
map_result_sorted = mapresult[order(sapply(mapresult,function(x){x$key}))]
# Step 2b: Make a table (using ldply, very slow)
mapresult_df  = data.frame(key = unlist(lapply(map_result_sorted, "[","key")),
	       		  value = unlist(lapply(map_result_sorted, "[","value")))

# Step 3: Apply the reducer over groups
reduce_result = ddply(mapresult_df, .(key), function(x) sum(x$value))
print(head(reduce_result));



# This is, how it works on Hadoop
# compare https://www.r-bloggers.com/from-functional-programming-to-mapreduce-in-r/

#word_count_hadoop <- function(lines) { 
#  words <- to.dfs(do.call(c, strsplit(clean(lines),'\s')))
#  fs.ptr <- mapreduce(input=words,
#    map=function(k,v) keyval(v,1),
#    reduce=function(k,v) keyval(k, sum(v)))
#  raw <- from.dfs(fs.ptr)
#  out <- raw$val
#  names(out) <- raw$key
#  out[order(names(out))]
#} 
